### PR TITLE
Fix embedding with sparse=True

### LINKED ISSUE #4666: nn.Embedding sparse=True not giving sparse gradient for master branch
I use the current master branch(see the code below) of PyTorch, and there seems a bug when I using sparse nn.Embedding:
when I set sparse=True for nn.Embedding, finally I still got dense gradient and here is the code:

```
>>> import torch
>>> from torch.autograd import Variable
>>> import torch.nn as nn

>>> emb = nn.Embedding(10,3, sparse=True)
>>> input = Variable(torch.LongTensor([[1],[2]]))
>>> loss = emb(input).mean()

>>> loss
Variable containing:
-0.3957
[torch.FloatTensor of size (1,)]

>>> loss.backward()

>>> emb.weight.grad
Variable containing:
 0.0000  0.0000  0.0000
 0.1667  0.1667  0.1667
 0.1667  0.1667  0.1667
 0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000
[torch.FloatTensor of size (10,3)]

>>> emb.weight.grad.is_sparse
False
```

`torch.__version__ is '0.4.0a0+8eded5a'`

However, the gradient is sparse when I change my PyTorch to version '0.3.0.post4':
```
>>> emb.weight.grad
Variable containing:FloatTensor of size 10x3 with indices:

 1  2
[torch.LongTensor of size 1x2]
and values:

 0.1667  0.1667  0.1667
 0.1667  0.1667  0.1667
[torch.FloatTensor of size 2x3]
```

### PR DESCRIPTION
**Pull request by colesbury**
Fixes #4666

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Review by gchanan — APPROVED** [review]
lgtm

**Comment by colesbury** [comment]
@ngimel yes! Thanks for the reminder. I'll add it in a subsequent PR
